### PR TITLE
utility: add benchmark library, but don't use it yet.

### PR DESCRIPTION
**Pull request by jmarantz**
*title*: *one line description*

This is a prereq for running CI on https://github.com/envoyproxy/envoy/pull/2348

*Description*:
Integrates github.com/google/benchmark so it can be used in future PRs for demonstrating improvement.  Note this PR does not actually depend on it, so it can pass CI.

*Risk Level*: Low

*Tested*:  //test/...

### CODE REVIEW

**Comment by jmarantz** [comment]
Note: I tried for several hours to get this to work with genrule_repository and raw genrule before giving up.  There don't appear to be any currently live examples of this flow working with a non-trivial C++ dependency, although @htuch pointed out there was one in the past: 

https://github.com/envoyproxy/envoy/blob/6f91542bdf626246968bd6e36f4d1e2417de4617/bazel/repositories.bzl#L317

But I had a hard time understanding how I would use this with an external github repo at a commit, as opposed to a versioned static file with a sha256.

I'm OK revisiting this later but right now this is something that IMO we need to fix a real perf issue.

**Comment by htuch** [comment]
@jmillikin-stripe can you provide an idea of what folks should be doing with `genrule_repository`? Alternatively, are you cool with use just using a build recipe here?

**Comment by jmillikin-stripe** [comment]
> But I had a hard time understanding how I would use this with an external github repo at a commit, as opposed to a versioned static file with a sha256.

`genrule_repository` doesn't know how to git clone, though we could teach it if needed. Using Github's archive tarballs would almost certainly be easier.

> Alternatively, are you cool with use just using a build recipe here?

I'm fine in general with using build recipes, but note that this particular recipe has some issues (e.g. it creates its own, non-hermetic googletest checkout).

What are the chances that this official Google benchmark framework could grow a build config for use with Google's official build system...?

**Comment by jmarantz** [comment]
The chance is pretty good we can get Bazel support: https://github.com/google/benchmark/pull/501

However I don't want to wait for that :)

I was able to get a good hint from one of the devs and managed to remove the gtest dependency.

**Comment by jmarantz** [comment]
RE 'archive tarballs' -- how do they work?  Is that something that has to be created by the repo owner?

**Comment by jmillikin-stripe** [comment]
See https://developer.github.com/v3/repos/contents/#get-archive-link

The link you'd want for google/benchmark is: https://github.com/google/benchmark/archive/e1c3a83b8197cf02e794f61228461c27d4e78cfb.tar.gz

File: ci/build_container/build_recipes/benchmark.sh [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+
+set -x
```
Comment (htuch) [inline ci/build_container/build_recipes/benchmark.sh]: Remove `-x` (once done debug)  and add `-e`.

File: ci/build_container/build_recipes/benchmark.sh [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+cd build
+cmake -G "Unix Makefiles" ../benchmark \
+  -DCMAKE_BUILD_TYPE=RELEASE \
+  -DBENCHMARK_ENABLE_GTEST_TESTS=OFF
+make
+cp src/libbenchmark.a "$THIRDPARTY_BUILD"/lib
+cd ../benchmark
+
+pwd
+include_dir="$THIRDPARTY_BUILD/include/testing/base/public"
```
Comment (htuch) [inline ci/build_container/build_recipes/benchmark.sh]: Prefer `INCLUDE_DIR` for consistency with other variables (I know in Google we use lowercase, but we would need to switch everything over).

File: ci/build_container/build_recipes/benchmark.sh [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+cd build
+cmake -G "Unix Makefiles" ../benchmark \
+  -DCMAKE_BUILD_TYPE=RELEASE \
+  -DBENCHMARK_ENABLE_GTEST_TESTS=OFF
+make
+cp src/libbenchmark.a "$THIRDPARTY_BUILD"/lib
+cd ../benchmark
+
+pwd
```
Comment (htuch) [inline ci/build_container/build_recipes/benchmark.sh]: Not needed.

File: ci/build_container/build_recipes/benchmark.sh (hunk 1 above)
Comment (jmarantz) [reply ci/build_container/build_recipes/benchmark.sh]: done

File: ci/build_container/build_recipes/benchmark.sh (hunk 3 above)
Comment (jmarantz) [reply ci/build_container/build_recipes/benchmark.sh]: done

File: ci/build_container/build_recipes/benchmark.sh (hunk 2 above)
Comment (jmarantz) [reply ci/build_container/build_recipes/benchmark.sh]: done; FWIW I thought lower-case variables were common in bash, and upper-case variables when they are exported.

Comment (htuch) [reply ci/build_container/build_recipes/benchmark.sh]: They are, and it's in the Google style guide, but we haven't had that convention in Envoy, so we tend to opt to be consistent with the existing code base as the lesser evil.

**Comment by htuch** [comment]
@jmarantz I think you may have skipped the push after update.

**Comment by jmarantz** [comment]
What's the next step for this PR...I am not (yet) a collaborator so I can't merge myself.

**Pull request merged**